rust/kernel/hw_random.rs

```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Hardware Random Number Generator.
+//!
+//! C header: [`include/linux/hw_random.h`](../../../../include/linux/hw_random.h)
+
+use alloc::{boxed::Box, slice::from_raw_parts_mut};
+
+use crate::{
+    bindings, c_types, error::from_kernel_result, str::CStr, types::PointerWrapper, Error, Result,
+};
+
+use core::{cell::UnsafeCell, marker::PhantomData, pin::Pin};
+
+/// This trait is implemented in order to provide callbacks to `struct hwrng`.
+pub trait HwrngOperations: Sized + 'static {
+    /// The methods to use to populate [`struct hwrng`].
+    const TO_USE: ToUse;
+
+    /// The pointer type that will be used to hold user-defined data type.
+    type Data: PointerWrapper = Box<Self>;
+
+    /// Initialization callback, can be left undefined.
+    fn init(_data: <Self::Data as PointerWrapper>::Borrowed<'_>) -> Result {
+        Err(Error::EINVAL)
+    }
+
+    /// Cleanup callback, can be left undefined.
+    fn cleanup(_data: <Self::Data as PointerWrapper>::Borrowed<'_>) {}
+
+    /// Read data into the provided buffer.
+    /// Drivers can fill up to max bytes of data into the buffer.
+    /// The buffer is aligned for any type and max is a multiple of 4 and >= 32 bytes.
+    fn read(
+        data: <Self::Data as PointerWrapper>::Borrowed<'_>,
+        buffer: &mut [i8],
+        wait: bool,
+    ) -> Result<i32>;
+}
+
+/// Registration structure for Hardware Random Number Generator driver.
+pub struct Registration<T: HwrngOperations> {
+    hwrng: UnsafeCell<bindings::hwrng>,
+    _p: PhantomData<T>,
+}
+
+impl<'a, T: HwrngOperations> Registration<T> {
```

Comment:
Why do we need this lifetime here?

Reply:
This was a lifetime of `&CStr` reference as shown in `new_pinned` function: `name: &'a CStr`.
It will be removed based on one of the solution provided below i.e. https://github.com/Rust-for-Linux/linux/pull/592#discussion_r767164031.